src/esphomelib/sensor/tx20_component.cpp

```diff
@@ -0,0 +1,160 @@
+    : PollingComponent(update_interval),
+      wind_speed_sensor_(new TX20WindSpeedSensor(wind_speed_name, this)),
+      wind_direction_sensor_(new TX20WindDirectionSensor(wind_direction_name, this)),
+      wind_direction_text_sensor_(new TX20WindDirectionTextSensor(wind_direction_text_name)),
+      pin_(pin) {
+  ESP_LOGV(TAG, "Start TX20Component...");
+  if(firstInstance){
+    ESP_LOGE(TAG, "Multiple TX20 Sensors are not supported!!");
+  }
+  firstInstance = true;
+}
+
+void TX20Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TX20...");
+  this->pin_->setup();
+
+  attachInterrupt(this->pin_->get_pin(), TX20Component::start_read_, RISING);
+  global_tx20_ = this;
+}
+void TX20Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TX20:");
+  LOG_PIN(" Pin:", TX20Component::pin_);
+  LOG_UPDATE_INTERVAL(this);
+}
+float TX20Component::get_setup_priority() const {
+  return setup_priority::HARDWARE_LATE;
+}
+
+void TX20Component::update() {
+  if (tx20_available) {
+    ESP_LOGV(TAG, "Updating TX20...");
+    std::string direction = DIRECTIONS.at(tx20_wind_direction);
+
+    this->wind_direction_sensor_->publish_state(tx20_wind_direction);
+    this->wind_speed_sensor_->publish_state(tx20_wind_speed_kmh);
+    this->wind_direction_text_sensor_->publish_state(direction);
+  }
+}
+void TX20Component::start_read_() {
+  global_tx20_->start_read_internal_();
+}
+
+void TX20Component::start_read_internal_() {
+  /* La Crosse TX20 Anemometer datagram every 2 seconds
+   * 0-0 11011 0011 111010101111 0101 1100 000101010000 0-0 - Received pin data
+   * at 1200 uSec per bit sa    sb   sc           sd   se   sf 00100 1100
+   * 000101010000 1010 1100 000101010000     - sa to sd inverted user data,
+   * LSB first sa - Start frame always 00100 sb - Wind direction 0 - 15 sc -
+   * Wind speed 0 - 511 sd - Checksum se - Wind direction 0 - 15 sf - Wind speed
+   * 0 - 511
+   */
+
+  tx20_available = false;
+
+  uint8_t tx20_sa = 0;
+  uint8_t tx20_sb = 0;
+  uint8_t tx20_sd = 0;
+  uint8_t tx20_se = 0;
+  uint8_t tx20_sc = 0;
+  uint8_t tx20_sf = 0;
+
+  delayMicroseconds(TX20_BIT_TIME / 2);
+
+  for (int bitcount = 41; bitcount > 0; bitcount--) {
+    uint8_t dpin = (digitalRead(this->pin_->get_pin()));
+    if (bitcount > 41 - 5) {
+      // start, inverted
+      tx20_sa = (tx20_sa << 1) | (dpin ^ 1);
+    } else if (bitcount > 41 - 5 - 4) {
+      // wind dir, inverted
+      tx20_sb = tx20_sb >> 1 | ((dpin ^ 1) << 3);
+    } else if (bitcount > 41 - 5 - 4 - 12) {
+      // windspeed, inverted
+      tx20_sc = tx20_sc >> 1 | ((dpin ^ 1) << 11);
+    } else if (bitcount > 41 - 5 - 4 - 12 - 4) {
+      // checksum, inverted
+      tx20_sd = tx20_sd >> 1 | ((dpin ^ 1) << 3);
+    } else if (bitcount > 41 - 5 - 4 - 12 - 4 - 4) {
+      // wind dir
+      tx20_se = tx20_se >> 1 | (dpin << 3);
+    } else {
+      // windspeed
+      tx20_sf = tx20_sf >> 1 | (dpin << 11);
+    }
+
+    delayMicroseconds(TX20_BIT_TIME);
```

Comment:
So this interrupt will take 41*1220 microseconds (50 milliseconds) to read the wind sensor? That's not good for a real-time system like an ESP - the WiFi controller will not be able to respond to packets during this time and esphomelib's core loop will not continue. Please try re-writing this into an asynchronous version, see the UART sensors for inspiration.

Reply:
Any one in particular? Did not quite work out the way as I was intending it to do:
- Intention was to move it to the loop and check the with micros() if the elapsed time is >= as my timeout. Result was not quite working.